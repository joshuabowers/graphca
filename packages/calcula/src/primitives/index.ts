import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
