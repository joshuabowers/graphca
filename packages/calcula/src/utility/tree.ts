import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
