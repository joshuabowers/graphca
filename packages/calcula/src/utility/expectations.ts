import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
