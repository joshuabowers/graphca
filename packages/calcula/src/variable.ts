import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
