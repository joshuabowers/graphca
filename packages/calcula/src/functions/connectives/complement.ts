import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation, isSpecies } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { Unary, unary, unaryFnRule, when } from "../../closures/unary"
import { Conjunction, and, andRule } from './conjunction'
import { Disjunction, or, orRule } from './disjunction'
import { ExclusiveDisjunction, xor, xorRule } from './exclusiveDisjunction'
import { Implication } from './implication'
import { AlternativeDenial, nand, nandRule } from './alternativeDenial'
import { JointDenial, nor, norRule } from './jointDenial'
import { Biconditional, xnor, xnorRule } from './biconditional'
import { ConverseImplication } from './converseImplication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Complement = Unary<Species.not, Genera.connective>

export const notRule = unaryFnRule(Unicode.not)

export const [not, isComplement, $not] = unary<Complement, Boolean>(
  Unicode.not, Notation.prefix, Species.not, Genera.connective
)(
  r => [boolean(r.value === 0), notRule(r), 'real complement'],
  c => [boolean(c.a === 0 && c.b === 0), notRule(c), 'complex complement'],
  b => [boolean(!b.value), notRule(b), 'boolean complement']
)(
  // NOTE: Cannot use derived isSpecies guards as they are not defined
  // by this point.
  when<Complement>(
    isSpecies(Species.not), 
    v => [v.expression, rule`${v.expression}`, 'double complement']
  ),
  when<Conjunction>(
    isSpecies(Species.and), 
    v => [nand(v.left, v.right), nandRule(v.left, v.right), 'complement of conjunction']
  ),
  when<Disjunction>(
    isSpecies(Species.or), 
    v => [nor(v.left, v.right), norRule(v.left, v.right), 'complement of disjunction']
  ),
  when<AlternativeDenial>(
    isSpecies(Species.nand), 
    v => [and(v.left, v.right), andRule(v.left, v.right), 'complement of alternative denial']
  ),
  when<JointDenial>(
    isSpecies(Species.nor), 
    v => [or(v.left, v.right), orRule(v.left, v.right), 'complement of joint denial']
  ),
  when<ExclusiveDisjunction>(
    isSpecies(Species.xor), 
    v => [xnor(v.left, v.right), xnorRule(v.left, v.right), 'complement of exclusive disjunction']
  ),
  when<Implication>(
    isSpecies(Species.implies),
    v => [
      and(v.left, not(v.right)), 
      rule`${v.left} ${Unicode.and} (${v.right})${Unicode.not}`, 
      'complement of implication'
    ]
  ),
  when<Biconditional>(
    isSpecies(Species.xnor),
    v => [xor(v.left, v.right), xorRule(v.left, v.right), 'complement of biconditional']
  ),
  when<ConverseImplication>(
    isSpecies(Species.converse),
    v => [
      and(not(v.left), v.right), 
      rule`(${v.left})${Unicode.not} ${Unicode.and} ${v.right}`,
      'complement of converse implication'
    ]
  )
)
