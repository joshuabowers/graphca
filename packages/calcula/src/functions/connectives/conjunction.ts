import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { isComplement } from './complement'
import { isDisjunction } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Conjunction = Connective<Species.and>

export const andRule = binaryInfixRule(Unicode.and)

export const [and, isConjunction, $and] = binary<Conjunction, Boolean>(
  Species.and, Genera.connective
)(
  (l, r) => [boolean(l.value !== 0 && r.value !== 0), andRule(l, r), 'real conjunction'],
  (l, r) => [
    boolean((l.a !== 0 || l.b !== 0) && (r.a !== 0 || r.b !== 0)), 
    andRule(l, r),
    'complex conjunction'
  ],
  (l, r) => [boolean(l.value && r.value), andRule(l, r), 'boolean conjunction']
)(
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive identity']
  ),
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive identity']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive idempotency']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  )
)
