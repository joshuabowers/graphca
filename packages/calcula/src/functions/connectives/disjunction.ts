import { _ } from '@arrows/multimethod'
import { TreeNode, Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement } from './complement'
import { isConjunction } from './conjunction'
import { implies } from './implication'
import { converse } from './converseImplication'
import { Unicode } from '../../Unicode'

export type Disjunction = Connective<Species.or>

export const [or, isDisjunction, $or] = binary<Disjunction, Boolean>(
  Unicode.or, Notation.infix, Species.or, Genera.connective
)(
  (l, r) => boolean(l.value.value !== 0 || r.value.value !== 0),
  (l, r) => boolean(
    l.value.a !== 0 || l.value.b !== 0 || r.value.a !== 0 || r.value.b !== 0
  ),
  (l, r) => boolean(l.value.value || r.value.value)
)(
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [l, 'disjunctive identity']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [r, 'disjunctive identity']
  ),
  when(
    [_, isValue(boolean(true))],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    [isValue(boolean(true)), _],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [l, 'disjunctive idempotency']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'disjunctive tautology']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'disjunctive tautology']
  ),
  when<Complement, TreeNode>(
    (l, _r) => isComplement(l),
    (l, r) => [
      implies(l.value.expression, r), 
      'disjunctive implication'
    ]
  ),
  when<TreeNode, Complement>(
    (_l, r) => isComplement(r),
    (l, r) => [
      converse(l, r.value.expression), 
      'disjunctive converse implication'
    ]
  )
)
