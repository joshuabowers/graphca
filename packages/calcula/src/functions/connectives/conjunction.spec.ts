import { expectWriter } from '../../utility/expectations'
import { Clades, Genera, Species } from '../../utility/tree'
import { real, complex, boolean } from '../../primitives'
import { variable } from '../../variable'
import { not } from './complement'
import { or } from './disjunction'
import { Conjunction, and } from './conjunction'

describe('and', () => {
  it('returns true when given two true things', () => {
    expectWriter(
      and(boolean(true), boolean(true))
    )(
      boolean(true),
      [[boolean(true), boolean(true)], 'conjunctive identity']
    )
  })

  it('returns false if the left argument is false', () => {
    expectWriter(
      and(boolean(false), boolean(true))
    )(
      boolean(false),
      [[boolean(false), boolean(true)], 'conjunctive identity']
    )
  })

  it('returns false if the right argument is false', () => {
    expectWriter(
      and(boolean(true), boolean(false))
    )(
      boolean(false),
      [[boolean(true), boolean(false)], 'conjunctive identity']
    )
  })

  it('returns false if both arguments are false', () => {
    expectWriter(
      and(boolean(false), boolean(false))
    )(
      boolean(false),
      [[boolean(false), boolean(false)], 'conjunctive annihilator']
    )
  })

  it('casts reals to booleans, where 0 is false, non-zero is true', () => {
    expectWriter(
      and(real(5), real(0))
    )(
      boolean(false),
      [[real(5), real(0)], 'real conjunction']
    )
  })

  it('casts complexes to booleans, 0 => false, non-0 => true', () => {
    expectWriter(
      and(complex([5,0]), complex([0,0]))
    )(
      boolean(false),
      [[complex([5, 0]), complex([0, 0])], 'complex conjunction']
    )
  })

  it('returns the left operand if the right is true', () => {
    expectWriter(
      and(variable('x'), boolean(true))
    )(
      variable('x'),
      [[variable('x'), boolean(true)], 'conjunctive identity']
    )
  })

  it('returns the right operand if the left is true', () => {
    expectWriter(
      and(boolean(true), variable('x'))
    )(
      variable('x'),
      [[boolean(true), variable('x')], 'conjunctive identity']
    )
  })

  it('returns the false if the right operand is false', () => {
    expectWriter(
      and(variable('x'), boolean(false))
    )(
      boolean(false),
      [[variable('x'), boolean(false)], 'conjunctive annihilator']
    )
  })

  it('returns false if the left operand is false', () => {
    expectWriter(
      and(boolean(false), variable('x'))
    )(
      boolean(false),
      [[boolean(false), variable('x')], 'conjunctive annihilator']
    )
  })

  it('returns the left operand if left equivalent to right', () => {
    expectWriter(
      and(variable('x'), variable('x'))
    )(
      variable('x'),
      [[variable('x'), variable('x')], 'conjunctive idempotency']
    )
  })

  // See (e.g.): https://en.wikipedia.org/wiki/Boolean_algebra#Monotone_laws
  // x | y | x or y | x and (x or y) | x and x | x and y | (x and x) or (x and y)
  // ----------------------------------------------------------------------------
  // F | F | F      | F              | F       | F       | F 
  // F | T | T      | F              | F       | F       | F
  // T | F | T      | T              | T       | F       | T
  // T | T | T      | T              | T       | T       | T
  it('returns the left operand if the right is a Disjunction of the left', () => {
    expectWriter(
      and(variable('x'), or(variable('x'), variable('y')))
    )(
      variable('x'),
      [[variable('x'), variable('y')], 'disjunction'],
      [[variable('x'), or(variable('x'), variable('y'))], 'conjunctive absorption']
    )
    expectWriter(
      and(variable('x'), or(variable('y'), variable('x')))
    )(
      variable('x'),
      [[variable('y'), variable('x')], 'disjunction'],
      [[variable('x'), or(variable('y'), variable('x'))], 'conjunctive absorption']
    )
  })

  it('returns the right operand if the left is a Disjunction of the right', () => {
    expectWriter(
      and(or(variable('x'), variable('y')), variable('x'))
    )(
      variable('x'),
      [[variable('x'), variable('y')], 'disjunction'],
      [[or(variable('x'), variable('y')), variable('x')], 'conjunctive absorption']
    )
    expectWriter(
      and(or(variable('y'), variable('x')), variable('x'))
    )(
      variable('x'),
      [[variable('y'), variable('x')], 'disjunction'],
      [[or(variable('y'), variable('x')), variable('x')], 'conjunctive absorption']
    )
  })

  it('returns false if the right operand is the negation of the left', () => {
    expectWriter(
      and(variable('x'), not(variable('x')))
    )(
      boolean(false),
      [variable('x'), 'complement'],
      [[variable('x'), not(variable('x'))], 'contradiction']
    )
  })

  it('returns false if the left operand is the negation of the right', () => {
    expectWriter(
      and(not(variable('x')), variable('x'))
    )(
      boolean(false),
      [variable('x'), 'complement'],
      [[not(variable('x')), variable('x')], 'contradiction']
    )
  })

  it('returns a Conjunction on variable input', () => {
    expectWriter(and(variable('x'), variable('y')))(
      {
        clade: Clades.binary, genus: Genera.connective, species: Species.and,
        left: variable('x'), right: variable('y')
      } as Conjunction,
      [[variable('x').value, variable('y').value], 'conjunction']
    )
  })
})
