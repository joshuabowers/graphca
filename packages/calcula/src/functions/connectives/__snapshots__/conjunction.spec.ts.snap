// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`and casts complexes to booleans, 0 => false, non-0 => true 1`] = `
[
  [
    "(5+0𝒊∧0+0𝒊)",
    "identified conjunction",
  ],
  [
    "([5+0𝒊]∧0+0𝒊)",
    "processing left operand",
  ],
  [
    "5+0𝒊",
    "created complex",
  ],
  [
    "({5+0𝒊}∧[0+0𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "(5+0𝒊∧{0+0𝒊})",
    "processed right operand",
  ],
  [
    "(5+0𝒊∧0+0𝒊)",
    "complex conjunction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and casts reals to booleans, where 0 is false, non-zero is true 1`] = `
[
  [
    "(5∧0)",
    "identified conjunction",
  ],
  [
    "([5]∧0)",
    "processing left operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}∧[0])",
    "processed left operand; processing right operand",
  ],
  [
    "0",
    "created real",
  ],
  [
    "(5∧{0})",
    "processed right operand",
  ],
  [
    "(5∧0)",
    "real conjunction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns a Conjunction on variable input 1`] = `
[
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
]
`;

exports[`and returns false if both arguments are false 1`] = `
[
  [
    "(false∧false)",
    "identified conjunction",
  ],
  [
    "([false]∧false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∧[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false∧{false})",
    "processed right operand",
  ],
  [
    "(false∧false)",
    "conjunctive annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns false if the left argument is false 1`] = `
[
  [
    "(false∧true)",
    "identified conjunction",
  ],
  [
    "([false]∧true)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∧[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(false∧{true})",
    "processed right operand",
  ],
  [
    "(false∧true)",
    "conjunctive identity",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns false if the left operand is false 1`] = `
[
  [
    "(false∧x)",
    "identified conjunction",
  ],
  [
    "([false]∧x)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(false∧{x})",
    "processed right operand",
  ],
  [
    "(false∧x)",
    "conjunctive annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns false if the left operand is the negation of the right 1`] = `
[
  [
    "(¬(x)∧x)",
    "identified conjunction",
  ],
  [
    "([¬(x)]∧x)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(¬(x)∧{x})",
    "processed right operand",
  ],
  [
    "(¬(x)∧x)",
    "conjunctive contradiction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns false if the right argument is false 1`] = `
[
  [
    "(true∧false)",
    "identified conjunction",
  ],
  [
    "([true]∧false)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∧[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(true∧{false})",
    "processed right operand",
  ],
  [
    "(true∧false)",
    "conjunctive identity",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns false if the right operand is the negation of the left 1`] = `
[
  [
    "(x∧¬(x))",
    "identified conjunction",
  ],
  [
    "([x]∧¬(x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[¬(x)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "(x∧{¬(x)})",
    "processed right operand",
  ],
  [
    "(x∧¬(x))",
    "conjunctive contradiction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns the false if the right operand is false 1`] = `
[
  [
    "(x∧false)",
    "identified conjunction",
  ],
  [
    "([x]∧false)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(x∧{false})",
    "processed right operand",
  ],
  [
    "(x∧false)",
    "conjunctive annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`and returns the left operand if left equivalent to right 1`] = `
[
  [
    "(x∧x)",
    "identified conjunction",
  ],
  [
    "([x]∧x)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(x∧{x})",
    "processed right operand",
  ],
  [
    "(x∧x)",
    "conjunctive idempotency",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the left operand if the right is a Disjunction of the left 1`] = `
[
  [
    "(x∧(x∨y))",
    "identified conjunction",
  ],
  [
    "([x]∧(x∨y))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[(x∨y)])",
    "processed left operand; processing right operand",
  ],
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
  [
    "(x∧{(x∨y)})",
    "processed right operand",
  ],
  [
    "(x∧(x∨y))",
    "conjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the left operand if the right is a Disjunction of the left 2`] = `
[
  [
    "(x∧(y∨x))",
    "identified conjunction",
  ],
  [
    "([x]∧(y∨x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[(y∨x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(y∨x)",
    "identified disjunction",
  ],
  [
    "([y]∨x)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(y∨{x})",
    "processed right operand",
  ],
  [
    "(y∨x)",
    "created disjunction",
  ],
  [
    "(x∧{(y∨x)})",
    "processed right operand",
  ],
  [
    "(x∧(y∨x))",
    "conjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the left operand if the right is true 1`] = `
[
  [
    "(x∧true)",
    "identified conjunction",
  ],
  [
    "([x]∧true)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(x∧{true})",
    "processed right operand",
  ],
  [
    "(x∧true)",
    "conjunctive identity",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the right operand if the left is a Disjunction of the right 1`] = `
[
  [
    "((x∨y)∧x)",
    "identified conjunction",
  ],
  [
    "([(x∨y)]∧x)",
    "processing left operand",
  ],
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
  [
    "({(x∨y)}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "((x∨y)∧{x})",
    "processed right operand",
  ],
  [
    "((x∨y)∧x)",
    "conjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the right operand if the left is a Disjunction of the right 2`] = `
[
  [
    "((y∨x)∧x)",
    "identified conjunction",
  ],
  [
    "([(y∨x)]∧x)",
    "processing left operand",
  ],
  [
    "(y∨x)",
    "identified disjunction",
  ],
  [
    "([y]∨x)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(y∨{x})",
    "processed right operand",
  ],
  [
    "(y∨x)",
    "created disjunction",
  ],
  [
    "({(y∨x)}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "((y∨x)∧{x})",
    "processed right operand",
  ],
  [
    "((y∨x)∧x)",
    "conjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns the right operand if the left is true 1`] = `
[
  [
    "(true∧x)",
    "identified conjunction",
  ],
  [
    "([true]∧x)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(true∧{x})",
    "processed right operand",
  ],
  [
    "(true∧x)",
    "conjunctive identity",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`and returns true when given two true things 1`] = `
[
  [
    "(true∧true)",
    "identified conjunction",
  ],
  [
    "([true]∧true)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∧[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(true∧{true})",
    "processed right operand",
  ],
  [
    "(true∧true)",
    "conjunctive identity",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
